### PR TITLE
session: api to resume canceled downloads from previous session

### PR DESCRIPTION
**Pull request by deepak1556**
- [x] Add spec

Fixes https://github.com/electron/electron/issues/7720

/cc @idododu

### CODE REVIEW

File: docs/api/download-item.md [hunk 1]
```diff
@@ -146,3 +146,23 @@ header.
 #### `downloadItem.getState()`
 
 Returns `String` - The current state.  Can be `progressing`, `completed`, `cancelled` or `interrupted`.
+
+**Note:** The following api are useful specifically to resume
```
Comment (zeke) [inline docs/api/download-item.md]: api -> APIs

or

api -> methods

Comment (zeke) [inline docs/api/download-item.md]: resume `cancelled` item -> resume a `cancelled` item

File: docs/api/download-item.md [hunk 2]
```diff
@@ -146,3 +146,23 @@ header.
...
+`cancelled` item when session is restarted.
+
+#### `downloadItem.getURLChain()`
+
+Returns `String[]` - The complete url chain of the item including any redirects.
+
+#### `downloadItem.getLastModifiedTime`
```
Comment (zeke) [inline docs/api/download-item.md]: These last three methods need parentheses, e.g. `getLastModifiedTime()`

File: docs/api/session.md [hunk 3]
```diff
@@ -344,6 +344,25 @@ Returns `String` - The user agent for this session.
...
+  * `mimeType` String (optional)
+  * `offset` Integer - Start range for the download.
+  * `length` Integer - Total length of the download.
+  * `lastModified` String - Last-Modified header value.
+  * `eTag` String - ETag header value.
+  * `startTime` Double (optional) - Time when download was started in
+    number of seconds since UNIX epoch.
+
+Allows resuming `cancelled` or `interrupted` downloads from previous `Session`.
+The api will generate [DownloadItem](download-item.md) that can be accessed with the [will-download](#event-will-download)
```
Comment (zeke) [inline docs/api/session.md]: api -> API

generate [DownloadItem] -> generate a [DownloadItem]

File: docs/api/session.md [hunk 4]
```diff
@@ -344,6 +344,25 @@ Returns `String` - The user agent for this session.
...
+event. The [DownloadItem](download-item.md) will not have any `WebContents` associated with it and
+the initial state will be `interrupted`. The download will start only when the
+`resume` api is called on the [DownloadItem](download-item.md).
```
Comment (zeke) [inline docs/api/session.md]: api -> API

**Comment by deepak1556** [comment]
@zeke Fixed, thanks!

File: spec/api-session-spec.js [hunk 5]
```diff
@@ -565,4 +582,84 @@ describe('session module', function () {
...
+        offset: 0,
+        length: 5242880
+      }
+      w.webContents.session.createInterruptedDownload(options)
+      ipcRenderer.once('download-created', function (event, state, urlChain,
+                                                     mimeType, receivedBytes,
+                                                     totalBytes, filename,
+                                                     savePath) {
+        assert.equal(state, 'interrupted')
+        assert.equal(urlChain, urlChain)
```
Comment (kevinsawicki) [inline spec/api-session-spec.js]: Should this be an `assert.deepEqual` since it is an array?

File: atom/browser/api/atom_api_session.cc [hunk 6]
```diff
@@ -520,6 +540,29 @@ void Session::GetBlobData(
...
+  options.Get("path", &path);
+  options.Get("urlChain", &url_chain);
+  options.Get("mimeType", &mime_type);
+  options.Get("offset", &offset);
+  options.Get("length", &length);
+  options.Get("lastModified", &last_modified);
+  options.Get("eTag", &etag);
+  options.Get("startTime", &start_time);
+  if (path.empty() || url_chain.empty() || length == 0 || offset >= length)
+    return;
```
Comment (kevinsawicki) [inline atom/browser/api/atom_api_session.cc]: Should this throw an error instead?

**Comment by kevinsawicki** [comment]
@deepak1556 this is really cool, left a few minor comments 👍

**Comment by deepak1556** [comment]
Have addressed the comments, thanks!

**Comment by kevinsawicki** [comment]
Thanks for this @deepak1556, rebased it and made one small spec tweak, 🚢 🚢 🚢

**Comment by thejhnz** [comment]
I am currently writing support for closing the application and for the download to finish when they re-open.

For some reason when resuming a "createInterruptedDownload" call it just doesn't work. canResume() returns false all the time. According to the API, the createInterruptedDownload function is meant to be able to be resumed?

I may be confused with what 'Resume' is, is it downloading from the start? I assumed that if you are downloading a 4gb file, download 1gb, application closes, do the createInterruptedDownload with the 1GB offset and it would continue. This is not the case at the moment.

If you guys could check it out to see if this is working correctly that would be great.

**Comment by deepak1556** [comment]
> canResume() returns false all the time.

This is something that needs to be fixed. You can check the state of the download item instead.

> I assumed that if you are downloading a 4gb file, download 1gb, application closes, do the createInterruptedDownload with the 1GB offset and it would continue. This is not the case at the moment.

Yes it should resume if you have provided the right parameters, have you tried resuming the download ?

**Comment by thejhnz** [comment]
Thanks for your reply.

I think am passing the correct parameters as it does send it to 'will-download'. I log the received bytes and it has to offset but once I resume it just changes the received bytes back to 0 and starts fresh. It does however create a new file, such as 'file (1).rar' and does not continue onto the other file 'file.rar'.

**Comment by thejhnz** [comment]
The adding of the (1) was something on my end. I have recoded it a little and now it is just overwriting the existing file, but starts off at 0 when resumed.

**Comment by thejhnz** [comment]
```
        if (item.getState() === 'interrupted') {
            console.log(item.getReceivedBytes());

            item.resume();

            console.log(item.getReceivedBytes());
        }
```

Results in :
```
Resuming Download from: 18632208
18632208
0
```

**Comment by deepak1556** [comment]
@thejhnz does the server responding file.rar support range requests ?

**Comment by thejhnz** [comment]
Yup
`HTTP/1.1 206 Partial Content
Date: Tue, 07 Mar 2017 07:50:39 GMT
Content-Type: application/x-rar-compressed
Content-Length: 61
Connection: close
Set-Cookie: __cfduid=d76d46be1b53f6d814ab1ef7e2d10df401488873039; expires=Wed, 07-Mar-18 07:50:39 GMT; path=/; domain=.xeragame.com; HttpOnly
Last-Modified: Wed, 04 Jan 2017 00:30:50 GMT
Accept-Ranges: bytes
Content-Range: bytes 60-120/4325188445
Server: cloudflare-nginx
CF-RAY: 33bbf7139450190e-AKL`

**Comment by deepak1556** [comment]
Thanks! can you provide a minimal example to reproduce the issue ? It will help to debug the problem.

**Comment by thejhnz** [comment]
No problem.

Not exactly a minimal example.. this is the electron-download-manager, I have made some modifications to check for existing file and if it does create the interupted download.

**Comment by HWRichardson** [comment]
Any update?

**Comment by thejhnz** [comment]
Any update on this?

**Comment by thejhnz** [comment]
I figured out that it **needs** the lastModified tag. Adding appears to have fixed the issue. It might be best to throw an error if this is not specified.

**Comment by deepak1556** [comment]
@thejhnz `lastModified` was not specified as optional in the docs, but I guess its better to throw error when not specified. Can you open an issue ? Also if you feel like submitting a PR , go ahead.

**Pull request merged**